File: windows/Beans.Core/BeansApiClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;

namespace Beans.Core;

public sealed class BeansApiClient(HttpClient httpClient)
{
    private readonly HttpClient _http = httpClient;
    private readonly SemaphoreSlim _refreshLock = new(1, 1);
    private TokenPair? _tokens;
    private Uri? _server;
    public event Action<TokenPair>? TokensChanged;

    public void Configure(Uri server) => _server = server;
    public void RestoreTokens(TokenPair tokens) => _tokens = tokens;
    public void ClearTokens() => _tokens = null;

    public Task<PendingRegistration> StartRegistrationAsync(string nickname, string email, byte[] authSecret, CryptoProfile profile, byte[] wrappedVaultKey, DeviceInput device, CancellationToken ct = default) =>
        SendAsync<PendingRegistration>(HttpMethod.Post, "v1/auth/register/start", new { nickname, email, authSecret = Convert.ToBase64String(authSecret), cryptoProfile = profile, wrappedVaultKey = Convert.ToBase64String(wrappedVaultKey), device }, false, ct);

    public Task<AuthSession> VerifyEmailAsync(Guid registrationId, string code, CancellationToken ct = default) =>
        SendAsync<AuthSession>(HttpMethod.Post, "v1/auth/verify-email", new { registrationId, code }, false, ct);

    public Task<AuthChallenge> LoginChallengeAsync(string email, CancellationToken ct = default) =>
        SendAsync<AuthChallenge>(HttpMethod.Post, "v1/auth/login/challenge", new { email }, false, ct);

    public Task<AuthSession> LoginAsync(string email, byte[] authSecret, DeviceInput device, CancellationToken ct = default) =>
        SendAsync<AuthSession>(HttpMethod.Post, "v1/auth/login", new { email, authSecret = Convert.ToBase64String(authSecret), device }, false, ct);

    public Task<VaultEnvelope> GetVaultAsync(CancellationToken ct = default) => SendAsync<VaultEnvelope>(HttpMethod.Get, "v1/vault/", null, true, ct);
    public Task PutVaultAsync(VaultEnvelope envelope, CancellationToken ct = default) => SendEmptyAsync(HttpMethod.Put, "v1/vault/", envelope, true, ct);
    public Task<SyncPage> PullSyncAsync(long cursor, CancellationToken ct = default) => SendAsync<SyncPage>(HttpMethod.Get, $"v1/sync/?cursor={cursor}", null, true, ct);
    public Task<IReadOnlyList<SyncEnvelope>> PushSyncAsync(IReadOnlyList<SyncEnvelope> records, CancellationToken ct = default) => SendAsync<IReadOnlyList<SyncEnvelope>>(HttpMethod.Post, "v1/sync/batch", records, true, ct);
    public Task<IReadOnlyList<DeviceRecord>> ListDevicesAsync(CancellationToken ct = default) => SendAsync<IReadOnlyList<DeviceRecord>>(HttpMethod.Get, "v1/devices/", null, true, ct);
    public Task RevokeDeviceAsync(Guid deviceId, CancellationToken ct = default) => SendEmptyAsync(HttpMethod.Delete, $"v1/devices/{deviceId}", null, true, ct);
    public Task LogoutAsync(CancellationToken ct = default) => SendEmptyAsync(HttpMethod.Post, "v1/auth/logout", null, true, ct);
    public async Task<PasswordResetStartResult> StartPasswordResetAsync(string email, CancellationToken ct = default)
    {
        using var response = await SendResponseAsync(HttpMethod.Post, "v1/auth/password/reset/start", new { email }, false, ct);
        var data = await response.Content.ReadAsByteArrayAsync(ct);
        return data.Length == 0
            ? new PasswordResetStartResult(null)
            : JsonSerializer.Deserialize<PasswordResetStartResult>(data, JsonOptions.Default) ?? new PasswordResetStartResult(null);
    }
    public Task CompletePasswordResetAsync(string email, string code, byte[] authSecret, CryptoProfile profile, byte[] wrappedVaultKey, CancellationToken ct = default) =>
        SendEmptyAsync(HttpMethod.Post, "v1/auth/password/reset/complete", new { email, code, authSecret = Convert.ToBase64String(authSecret), cryptoProfile = profile, wrappedVaultKey = Convert.ToBase64String(wrappedVaultKey) }, false, ct);

    public Task<QrSession> CreateQrSessionAsync(DeviceInput device, byte[] publicKey, byte[] exchangeSecret, CancellationToken ct = default) =>
        SendAsync<QrSession>(HttpMethod.Post, "v1/qr-sessions/", new { device, ephemeralPublicKey = Convert.ToBase64String(publicKey), exchangeSecret = Convert.ToBase64String(exchangeSecret) }, false, ct);
    public Task<QrSession> GetQrSessionAsync(Guid id, CancellationToken ct = default) => SendAsync<QrSession>(HttpMethod.Get, $"v1/qr-sessions/{id}", null, false, ct);
    public Task ApproveQrSessionAsync(Guid id, string verificationCode, byte[] encryptedVaultKey, CancellationToken ct = default) =>
        SendEmptyAsync(HttpMethod.Post, $"v1/qr-sessions/{id}/approve", new { verificationCode, encryptedVaultKey = Convert.ToBase64String(encryptedVaultKey) }, true, ct);
    public Task<QrExchangeResult> ExchangeQrSessionAsync(QrLoginContext context, DeviceInput device, CancellationToken ct = default) =>
        SendAsync<QrExchangeResult>(HttpMethod.Post, $"v1/qr-sessions/{context.Session.Id}/exchange", new { device, ephemeralPublicKey = Convert.ToBase64String(AccountCrypto.PublicKeyFromPrivate(context.PrivateKey)), exchangeSecret = Convert.ToBase64String(context.ExchangeSecret) }, false, ct);

    private async Task<T> SendAsync<T>(HttpMethod method, string path, object? body, bool authorized, CancellationToken ct, bool retried = false)
    {
        using var request = CreateRequest(method, path);
        if (body is not null) request.Content = JsonContent.Create(body, options: JsonOptions.Default);
        if (authorized && _tokens is not null) request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _tokens.AccessToken);
        using var response = await _http.SendAsync(request, ct);
        if (response.StatusCode == HttpStatusCode.Unauthorized && authorized && !retried && await RefreshAsync(ct))
            return await SendAsync<T>(method, path, body, true, ct, true);
        await EnsureSuccessAsync(response, ct);
        return await response.Content.ReadFromJsonAsync<T>(JsonOptions.Default, ct) ?? throw new InvalidDataException("服务器返回为空");
    }

    private async Task SendEmptyAsync(HttpMethod method, string path, object? body, bool authorized, CancellationToken ct, bool retried = false)
    {
        using var request = CreateRequest(method, path);
        if (body is not null) request.Content = JsonContent.Create(body, options: JsonOptions.Default);
        if (authorized && _tokens is not null) request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _tokens.AccessToken);
        using var response = await _http.SendAsync(request, ct);
        if (response.StatusCode == HttpStatusCode.Unauthorized && authorized && !retried && await RefreshAsync(ct))
        {
            await SendEmptyAsync(method, path, body, true, ct, true);
            return;
        }
        await EnsureSuccessAsync(response, ct);
    }

    private async Task<HttpResponseMessage> SendResponseAsync(HttpMethod method, string path, object? body, bool authorized, CancellationToken ct, bool retried = false)
    {
        using var request = CreateRequest(method, path);
        if (body is not null) request.Content = JsonContent.Create(body, options: JsonOptions.Default);
        if (authorized && _tokens is not null) request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _tokens.AccessToken);
        var response = await _http.SendAsync(request, ct);
        if (response.StatusCode == HttpStatusCode.Unauthorized && authorized && !retried && await RefreshAsync(ct))
        {
            response.Dispose();
            return await SendResponseAsync(method, path, body, true, ct, true);
        }
        try { await EnsureSuccessAsync(response, ct); return response; }
        catch { response.Dispose(); throw; }
    }

    private async Task<bool> RefreshAsync(CancellationToken ct)
    {
        if (_tokens is null) return false;
        await _refreshLock.WaitAsync(ct);
        try
        {
            var refresh = _tokens.RefreshToken;
            _tokens = await SendAsync<TokenPair>(HttpMethod.Post, "v1/auth/refresh", new { refreshToken = refresh }, false, ct);
            TokensChanged?.Invoke(_tokens);
            return true;
        }
        finally { _refreshLock.Release(); }
    }

    private HttpRequestMessage CreateRequest(HttpMethod method, string path) =>
        new(method, new Uri(_server ?? throw new InvalidOperationException("尚未配置 Beans 服务地址"), path));

    private static async Task EnsureSuccessAsync(HttpResponseMessage response, CancellationToken ct)
    {
        if (response.IsSuccessStatusCode) return;
        var problem = await response.Content.ReadAsStringAsync(ct);
        throw new HttpRequestException($"Beans 服务请求失败 ({(int)response.StatusCode})：{problem}", null, response.StatusCode);
    }
}

public static class JsonOptions
{
    public static readonly JsonSerializerOptions Default = new(JsonSerializerDefaults.Web)
    {
        PropertyNameCaseInsensitive = true
    };
}

File: Beans/BeansAPIClient.swift
import Foundation

actor BeansAPIClient {
    static let shared = BeansAPIClient()

    private var baseURL: URL?
    private var tokens: BeansTokenPair?
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    private init() {
        encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        if let data = BeansSecureStore.shared.data(for: BeansSecureKey.accessToken),
           let saved = try? decoder.decode(BeansTokenPair.self, from: data) {
            tokens = saved
        }
    }

    func configure(serverURL: String) throws {
        let trimmed = serverURL.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = URL(string: trimmed), let scheme = url.scheme?.lowercased(), ["http", "https"].contains(scheme), url.host != nil else {
            throw BeansAPIError.invalidServerURL
        }
#if os(iOS) && !targetEnvironment(simulator) && !targetEnvironment(macCatalyst)
        let host = url.host?.lowercased() ?? ""
        if host == "localhost" || host == "127.0.0.1" || host == "::1" {
            throw BeansAPIError.loopbackUnavailable
        }
#endif
        baseURL = url
    }

    func restoreTokens(_ value: BeansTokenPair) {
        tokens = value
        if let data = try? encoder.encode(value) { _ = BeansSecureStore.shared.set(data, for: BeansSecureKey.accessToken) }
    }

    func clearTokens() {
        tokens = nil
        BeansSecureStore.shared.remove(BeansSecureKey.accessToken)
        BeansSecureStore.shared.remove(BeansSecureKey.refreshToken)
    }

    func startRegistration(nickname: String, email: String, authSecret: Data, profile: BeansCryptoProfile, wrappedVaultKey: Data, device: BeansDeviceInput) async throws -> BeansPendingRegistration {
        struct Body: Encodable { let nickname: String; let email: String; let authSecret: String; let cryptoProfile: BeansCryptoProfile; let wrappedVaultKey: String; let device: BeansDeviceInput }
        return try await send("/v1/auth/register/start", method: "POST", body: Body(nickname: nickname, email: email, authSecret: authSecret.base64EncodedString(), cryptoProfile: profile, wrappedVaultKey: wrappedVaultKey.base64EncodedString(), device: device), authorized: false)
    }

    func verifyEmail(registrationID: UUID, code: String) async throws -> BeansAuthSession {
        struct Body: Encodable { let registrationId: UUID; let code: String }
        return try await send("/v1/auth/verify-email", method: "POST", body: Body(registrationId: registrationID, code: code), authorized: false)
    }

    func loginChallenge(email: String) async throws -> BeansAuthChallenge {
        struct Body: Encodable { let email: String }
        return try await send("/v1/auth/login/challenge", method: "POST", body: Body(email: email), authorized: false)
    }

    func login(email: String, authSecret: Data, device: BeansDeviceInput) async throws -> BeansAuthSession {
        struct Body: Encodable { let email: String; let authSecret: String; let device: BeansDeviceInput }
        return try await send("/v1/auth/login", method: "POST", body: Body(email: email, authSecret: authSecret.base64EncodedString(), device: device), authorized: false)
    }

    func logout() async { _ = try? await sendEmpty("/v1/auth/logout", method: "POST", authorized: true) }

    func getVault() async throws -> BeansVaultEnvelope {
        try await send("/v1/vault/", method: "GET", authorized: true)
    }

    func putVault(_ value: BeansVaultEnvelope) async throws {
        _ = try await sendEmpty("/v1/vault/", method: "PUT", body: value, authorized: true)
    }

    func createQRSession(device: BeansDeviceInput, publicKey: Data, exchangeSecret: Data) async throws -> BeansQRSession {
        struct Body: Encodable { let device: BeansDeviceInput; let ephemeralPublicKey: String; let exchangeSecret: String }
        return try await send("/v1/qr-sessions/", method: "POST", body: Body(device: device, ephemeralPublicKey: publicKey.base64EncodedString(), exchangeSecret: exchangeSecret.base64EncodedString()), authorized: false)
    }

    func getQRSession(_ id: UUID) async throws -> BeansQRSession {
        try await send("/v1/qr-sessions/\(id)", method: "GET", authorized: false)
    }

    func approveQRSession(_ id: UUID, code: String, encryptedVaultKey: Data) async throws {
        struct Body: Encodable { let verificationCode: String; let encryptedVaultKey: String }
        _ = try await sendEmpty("/v1/qr-sessions/\(id)/approve", method: "POST", body: Body(verificationCode: code, encryptedVaultKey: encryptedVaultKey.base64EncodedString()), authorized: true)
    }

    func exchangeQRSession(_ context: BeansQRLoginContext, device: BeansDeviceInput) async throws -> BeansQRExchangeResult {
        struct Body: Encodable { let device: BeansDeviceInput; let ephemeralPublicKey: String; let exchangeSecret: String }
        let publicKey = try BeansAccountCrypto.makePublicKey(privateKey: context.privateKey)
        return try await send("/v1/qr-sessions/\(context.session.id)/exchange", method: "POST", body: Body(device: device, ephemeralPublicKey: publicKey.base64EncodedString(), exchangeSecret: context.exchangeSecret.base64EncodedString()), authorized: false)
    }

    func pullSync(cursor: Int64) async throws -> BeansSyncPage {
        try await send("/v1/sync/?cursor=\(cursor)", method: "GET", authorized: true)
    }

    func listDevices() async throws -> [BeansDeviceRecord] {
        try await send("/v1/devices/", method: "GET", authorized: true)
    }

    func revokeDevice(_ id: UUID) async throws {
        _ = try await sendEmpty("/v1/devices/\(id)", method: "DELETE", authorized: true)
    }

    func startPasswordReset(email: String) async throws -> String? {
        struct Body: Encodable { let email: String }
        struct Response: Decodable { let developmentCode: String? }
        let body = try encoder.encode(Body(email: email))
        let (data, response) = try await execute("/v1/auth/password/reset/start", method: "POST", bodyData: body, authorized: false)
        guard (200..<300).contains(response.statusCode) else { throw decodeError(data, status: response.statusCode) }
        guard !data.isEmpty else { return nil }
        return try? decoder.decode(Response.self, from: data).developmentCode
    }

    func completePasswordReset(email: String, code: String, authSecret: Data, profile: BeansCryptoProfile, wrappedVaultKey: Data) async throws {
        struct Body: Encodable { let email: String; let code: String; let authSecret: String; let cryptoProfile: BeansCryptoProfile; let wrappedVaultKey: String }
        let body = Body(email: email, code: code, authSecret: authSecret.base64EncodedString(), cryptoProfile: profile, wrappedVaultKey: wrappedVaultKey.base64EncodedString())
        _ = try await sendEmpty("/v1/auth/password/reset/complete", method: "POST", body: body, authorized: false)
    }

    func pushSync(_ records: [BeansSyncEnvelope]) async throws -> [BeansSyncEnvelope] {
        try await send("/v1/sync/batch", method: "POST", body: records, authorized: true)
    }

    private func send<Response: Decodable>(_ path: String, method: String, authorized: Bool, retrying: Bool = false) async throws -> Response {
        try await send(path, method: method, bodyData: nil, authorized: authorized, retrying: retrying)
    }

    private func send<Body: Encodable, Response: Decodable>(_ path: String, method: String, body: Body, authorized: Bool, retrying: Bool = false) async throws -> Response {
        try await send(path, method: method, bodyData: try encoder.encode(body), authorized: authorized, retrying: retrying)
    }

    private func send<Response: Decodable>(_ path: String, method: String, bodyData: Data?, authorized: Bool, retrying: Bool) async throws -> Response {
        let (data, response) = try await execute(path, method: method, bodyData: bodyData, authorized: authorized)
        if response.statusCode == 401, authorized, !retrying, try await refreshIfPossible() {
            return try await send(path, method: method, bodyData: bodyData, authorized: true, retrying: true)
        }
        guard (200..<300).contains(response.statusCode) else { throw decodeError(data, status: response.statusCode) }
        do { return try decoder.decode(Response.self, from: data) }
        catch { throw BeansAPIError.invalidResponse }
    }

    private func sendEmpty<Body: Encodable>(_ path: String, method: String, body: Body, authorized: Bool) async throws -> Bool {
        try await sendEmpty(path, method: method, bodyData: try encoder.encode(body), authorized: authorized)
    }

    private func sendEmpty(_ path: String, method: String, authorized: Bool) async throws -> Bool {
        try await sendEmpty(path, method: method, bodyData: nil, authorized: authorized)
    }

    private func sendEmpty(_ path: String, method: String, bodyData: Data?, authorized: Bool, retrying: Bool = false) async throws -> Bool {
        let (data, response) = try await execute(path, method: method, bodyData: bodyData, authorized: authorized)
        if response.statusCode == 401, authorized, !retrying, try await refreshIfPossible() {
            return try await sendEmpty(path, method: method, bodyData: bodyData, authorized: true, retrying: true)
        }
        guard (200..<300).contains(response.statusCode) else { throw decodeError(data, status: response.statusCode) }
        return true
    }

    private func execute(_ path: String, method: String, bodyData: Data?, authorized: Bool) async throws -> (Data, HTTPURLResponse) {
        guard let baseURL, let url = URL(string: path, relativeTo: baseURL) else { throw BeansAPIError.invalidServerURL }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 25
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let bodyData { request.httpBody = bodyData; request.setValue("application/json", forHTTPHeaderField: "Content-Type") }
        if authorized, let token = tokens?.accessToken { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch let error as URLError {
            BeansLogger.shared.log("Beans 账号服务连接失败：code=\(error.code.rawValue)", level: .error)
            throw BeansAPIError.network
        } catch {
            throw BeansAPIError.network
        }
        guard let http = response as? HTTPURLResponse else { throw BeansAPIError.network }
        return (data, http)
    }

    private func refreshIfPossible() async throws -> Bool {
        guard let refresh = tokens?.refreshToken else { return false }
        struct Body: Encodable { let refreshToken: String }
        let pair: BeansTokenPair = try await send("/v1/auth/refresh", method: "POST", body: Body(refreshToken: refresh), authorized: false)
        restoreTokens(pair)
        return true
    }

    private func decodeError(_ data: Data, status: Int) -> Error {
        struct Problem: Decodable { let detail: String?; let title: String? }
        let problem = try? decoder.decode(Problem.self, from: data)
        return BeansAPIError.server(status, problem?.detail ?? problem?.title ?? "服务器请求失败")
    }
}

enum BeansAPIError: LocalizedError {
    case invalidServerURL, loopbackUnavailable, network, invalidResponse, server(Int, String)

    var errorDescription: String? {
        switch self {
        case .invalidServerURL: return "请先配置有效的 Beans 账号服务地址"
        case .loopbackUnavailable: return "真机不能使用 localhost，请填写电脑的局域网地址或已部署的公网服务地址"
        case .network: return "无法连接 Beans 账号服务，请检查服务地址、网络和后端是否已启动"
        case .invalidResponse: return "Beans 账号服务返回了无法识别的数据"
        case .server(_, let message): return message
        }
    }
}
